File: python/637. 二叉树的层平均值.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

from typing import List

class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        if not root:
            return []
        from collections import deque
        queue = deque()
        queue.append(root)
        res = []
        while queue:
            tmp = []
            for _ in range(len(queue)):
                node = queue.popleft()
                tmp.append(node.val)
                if node.left: queue.append(node.left)
                if node.right: queue.append(node.right)
            res.append(sum(tmp)*1.0/len(tmp))
        return res


# dfs

class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        import collections
        traverse = collections.defaultdict(list)
        def dfs(node, n):
            if node:
                traverse[n] += [node.val]
                dfs(node.left,n+1)
                dfs(node.right,n+1)
        dfs(root, 0)
        return [sum(vals)/len(vals) for vals in traverse.values()]

File: go/general/637. 二叉树的层平均值.go
package main

func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return nil
	}
	var queue []*TreeNode
	queue = append(queue, root)
	var res []float64
	for len(queue) != 0 {
		var tmp []int
		l := len(queue)
		for i := 0; i < l; i++ {
			node := queue[0]
			tmp = append(tmp, node.Val)
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		res = append(res, float64(sum(tmp))/float64(len(tmp)))
	}
	return res
}
func sum(t []int) int {
	res := 0
	for _, v := range t {
		res += v
	}
	return res
}
